File: Python/Merge_Intervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    # @param intervals, a list of Interval
    # @return a list of Interval
    def merge(self, intervals):
        if len(intervals) <= 1: return intervals
        intervals.sort(key = lambda x: x.start)
        ret = []
        ret.append(intervals[0])
        n = len(intervals)
        for i in range(1, n):
            if ret[-1].end >= intervals[i].start:
                ret[-1].end = max(ret[-1].end, intervals[i].end)
            else:
                ret.append(intervals[i])
        return ret

File: Go/Merge_Intervals.go
/**
 * Definition for an interval.
 * type Interval struct {
 *	   Start int
 *	   End   int
 * }
 */
type Intervals []Interval

func (slice Intervals) Len() int {
    return len(slice)
}
func (slice Intervals) Less(x,y int) bool {
    return slice[x].Start < slice[y].Start
}
func (slice Intervals) Swap(x,y int) {
    slice[x],slice[y] = slice[y],slice[x]
}

func merge(intervals []Interval) []Interval {
    ret := []Interval{}
    if len(intervals) == 0 { return ret }
    sort.Sort(Intervals(intervals))
    cand := intervals[0]
    for i := 1; i < len(intervals); i++ {
        if intervals[i].Start > cand.End {
            ret = append(ret, cand)
            cand = intervals[i]
        } else {
            if cand.End < intervals[i].End {
                cand.End = intervals[i].End
            }
        }
    }
    ret = append(ret, cand)
    return ret
}
